### PR TITLE
Lua memory (de)allocation tracer:

### PR DESCRIPTION
**Pull request by projectkk2glider**
Usage:
 * turn on: `cmake -DLUA=YES -DLUA_ALLOCATOR_TRACER=YES -DDEBUG=YES -DNANO=NO`
 * get debug output and make plot data: `./simu 2>&1 | grep "^LT" | ../radio/util/lua_trace2plot.py > data.plot`
 * plot: `gnuplot  -e 'set xtics rotate;  plot "data.plot" using 2:xtic(1) ; pause mouse close'`

This is related to #4558 and should eventually be extended to be able to display results in a Companion after the simulator was run. 

With default options (-DLUA_ALLOCATOR_TRACER=NO) it should not affect anything.

With this enabled we get this output:
```
luaInitThemesAndWidgets
lsWidgets 0xd00ff608
luaLoadFiles() /THEMES
luaLoadFile(/THEMES/TBS/main.lua)
luaLoadScriptFileToState(/THEMES/TBS/main.lua, T): loading /THEMES/TBS/main.lua
LT: [+13735,-3994] lua_newstate(widgets):0
LT: [+48,-0] @/THEMES/TBS/main.lua:31
LT: [+24,-0] @/THEMES/TBS/main.lua:37
LT: [+20,-0] @/THEMES/TBS/main.lua:50
luaLoadThemeCallback()
luaLoadFiles() /WIDGETS
luaLoadFile(/WIDGETS/Counter/main.lua)
luaLoadScriptFileToState(/WIDGETS/Counter/main.lua, T): loading /WIDGETS/Counter/main.lua
LT: [+6318,-2279] @/THEMES/TBS/main.lua:52
LT: [+118,-0] @/WIDGETS/Counter/main.lua:1
LT: [+59,-0] @/WIDGETS/Counter/main.lua:2
LT: [+59,-0] @/WIDGETS/Counter/main.lua:3
LT: [+20,-0] @/WIDGETS/Counter/main.lua:11
LT: [+16,-0] @/WIDGETS/Counter/main.lua:15
LT: [+16,-0] @/WIDGETS/Counter/main.lua:19
LT: [+20,-0] @/WIDGETS/Counter/main.lua:24
LT: [+44,-22] @/WIDGETS/Counter/main.lua:21
```

The lines beginning with `LT: ` are the result of this PR and are used to debug Lua memory allocations with a resolution of one script line. 

The included python script can process this data and produce a nice graph like this:
screenshot

### CODE REVIEW

**Comment by powercycle** [comment]
Does the graph depicted show a memory leak?  I guess maybe it is also showing how many widgets are allocated and how much memory is taken?  Then it shows the deallocation of those items.  Nice tool.

**Comment by projectkk2glider** [comment]
>Does the graph depicted show a memory leak?

The graphs is just a part of the whole execution timeline. In reality the memory is freed at the end (if there are no memory leaks, which I haven't seen yet).

The tool basically counts how much memory was allocated and how much was freed for each script line.  So for one line there might be 10 allocations of various sizes and 5 frees again of various sizes. The tool will report the sum of 10 allocations and sum of 5 frees. For  example:
```
LT: [+44,-22] @/WIDGETS/Counter/main.lua:21
```
This means:
 * on line 21 in  /WIDGETS/Counter/main.lua
 * there were totaly 44 bytes allocated (it might be one or many small allocations, sum is 44 bytes)
 * and 22 bytes freed (it might be one or many small de-allocatons, sum is 22 bytes)

The way it is done now is a compromise, I would be possible to output each individual allocation/free, but  we would hit the limit of the logging speed and information would be lost.

**Comment by powercycle** [comment]
This seems reasonable.  Just knowing is one step forward from my view.

**Comment by pau33** [comment]
I am still on 2.1.9 yet, as everything ist fine besides the memory leak from time to time. I will update in Oktober, when the gilder season is over. During summer I sometimes have (lately rarely) the errors, I told before (when switching between models). There is (for me) no way to definitely produce them. Now I have not only 2 models, that I switch between, but 5 models. The errors do not occur more often.

**Comment by driver33** [comment]
02.09.2017 12:23 PM "Damjan Adamic" <notifications@github.com> napisał(a):

File: radio/src/lua/interface.cpp [hunk 1]
```diff
@@ -64,19 +97,38 @@ int custom_lua_atpanic(lua_State * L)
...
-    // From now on, as soon as a line is executed, error
-    // keep erroring until you're script reaches the top
-    lua_sethook(L, luaHook, LUA_MASKLINE, 0);
-    luaL_error(L, "CPU limit");
+  if (ar->event == LUA_HOOKCOUNT) {
+    instructionsPercent++;
+    if (instructionsPercent > 100) {
+      // From now on, as soon as a line is executed, error
+      // keep erroring until you're script reaches the top
+      // lua_sethook(L, luaHook, LUA_MASKLINE, 0);
```
Comment (bsongis) [inline radio/src/lua/interface.cpp]: why did you comment out this line?

File: radio/util/lua_trace2plot.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                memUsed += alloc
+                print("'%s'\t%d" % (line, memUsed))
+                x += 1
+            if free < 0:
+                memUsed += free
+                print("'%s'\t%d" % (line, memUsed))
+                x += 1
+
+inp.close()
+sys.exit(0)
```
Comment (bsongis) [inline radio/util/lua_trace2plot.py]: not needed

**Review by bsongis — APPROVED** [review]
Excellent PR!

File: radio/src/lua/interface.cpp (hunk 1 above)
Comment (projectkk2glider) [reply radio/src/lua/interface.cpp]: I forgot to enable it back(I commented it out for testing). I will enable it back and test...

File: radio/util/lua_trace2plot.py (hunk 2 above)
Comment (projectkk2glider) [reply radio/util/lua_trace2plot.py]: True

**Comment by projectkk2glider** [comment]
Tested with `-DLUA_ALLOCATOR_TRACER=NO` and `-DLUA_ALLOCATOR_TRACER=YES` on Horus, appears to work.

**Pull request merged**